cpp/src/round/round.cu

```diff
@@ -47,102 +52,190 @@ T __device__ generic_abs(T value)
+                    input.begin<T>(),
+                    input.end<T>(),
+                    out_view.begin<T>(),
+                    RoundFunctor{n});
+
+  return result;
+}
+
+struct round_type_dispatcher {
+  template <typename T, typename... Args>
+  std::enable_if_t<not is_supported_round_type<T>(), std::unique_ptr<column>> operator()(
+    Args&&... args)
+  {
+    CUDF_FAIL("Type not support for cudf::round");
   }
 
   template <typename T>
-  std::enable_if_t<std::is_integral<T>::value, std::unique_ptr<column>> operator()(
+  std::enable_if_t<is_supported_round_type<T>(), std::unique_ptr<column>> operator()(
     column_view const& input,
     int32_t decimal_places,
     cudf::rounding_method method,
     cudaStream_t stream,
     rmm::mr::device_memory_resource* mr)
   {
-    // only need to handle the case where decimal_places is < zero
-    // integers by definition have no fractional part, so result of "rounding" is a no-op
-    if (decimal_places >= 0) return std::make_unique<cudf::column>(input, stream, mr);
-
-    auto result = cudf::make_fixed_width_column(input.type(),  //
-                                                input.size(),
-                                                copy_bitmask(input, stream, mr),
-                                                input.null_count(),
-                                                stream,
-                                                mr);
-
-    auto out_view = result->mutable_view();
-    auto const n  = static_cast<T>(std::pow(10, -decimal_places));
-
-    thrust::transform(rmm::exec_policy(stream)->on(stream),
-                      input.begin<T>(),
-                      input.end<T>(),
-                      out_view.begin<T>(),
-                      [n] __device__(T e) -> T {
-                        auto const down = (e / n) * n;  // result from rounding down
-                        auto const sign = is_negative(e) ? -1 : 1;
-                        return down + sign * (generic_abs(e - down) >= n / 2 ? n : 0);
-                      });
-
-    return result;
+    // clang-format off
+    switch (method) {
+      case cudf::rounding_method::HALF_UP:
+        if      (decimal_places == 0) return round_with<T, half_up_zero    <T>>(input, decimal_places, stream, mr);
+        else if (decimal_places >  0) return round_with<T, half_up_positive<T>>(input, decimal_places, stream, mr);
+        else                          return round_with<T, half_up_negative<T>>(input, decimal_places, stream, mr);
+      case cudf::rounding_method::HALF_EVEN:
+        if      (decimal_places == 0) return round_with<T, half_even_zero    <T>>(input, decimal_places, stream, mr);
+        else if (decimal_places >  0) return round_with<T, half_even_positive<T>>(input, decimal_places, stream, mr);
+        else                          return round_with<T, half_even_negative<T>>(input, decimal_places, stream, mr);
+      default: CUDF_FAIL("Undefined rounding method");
+    }
```

Comment:
So this is beautiful.
 It is too bad that `T` is in there twice. This means someone could technically call `round_with` with conflicting types: `round_with<int32, half_up_zero<double>>`. I wonder if there is a way to extract the `T` type from the functor type inside of `round_with()` function. Something like `using T = decltype(RoundFunctor::n);`  Since all the functors have a variable `T n`.
Either way, this looks great.

Reply:
See https://github.com/rapidsai/cudf/pull/6647/files#r517540005. I don't think the `half_up_zero`, etc. even needs to be a class template and would avoid any conflicting types.

Reply:
Alternatively, if `half_up_zero` and the like need to be template classes, then they could be passed to `round_with` as a template-template argument:
```
template <typename T, template <typename> typename RoundFunctor>
std::unique_ptr<column> round_with(...){
  using Functor = RoundFunctor<T>;
}
...

round_with<T, half_up_zero>

Reply:
Pushed. The template template is awesome!